### ISSUE TITLE
AdventureNBTSerialization throw IllegalStateException sometimes using SystemChat wrapper

### ISSUE THREAD

**Issue by rayanbzd** [retrooper/packetevents#673]
**Describe the bug**

**Software brand**
Velocity 3.3.0-SNAPSHOT Build 346
PacketEvents 2.2.0-SNAPSHOT (dev build 323, latest at the moment)

**Plugins**
Just a custom plugin that use the code below

**How To Reproduce**
Create a plugin that use WrapperPlayServerSystemChatMessage and try to send a system chat message with multiple font, hover or click events or append multiple component togethers. 
I can't figure out what exactly make it throw the exception. For simple message without hover, clickevent or things like that, it work properly. But when you start to use complex component this problem appear.
```
    @Override
    public void onPacketSend(PacketSendEvent event) {
        Player player = (Player) event.getPlayer();
        if(event.getConnectionState() != ConnectionState.PLAY)return;
        if(event.getUser().getClientVersion().isOlderThan(ClientVersion.V_1_20_2))return;
        if (event.getPacketType() == PacketType.Play.Server.SYSTEM_CHAT_MESSAGE) {
            WrapperPlayServerSystemChatMessage wrapper = new WrapperPlayServerSystemChatMessage(event);
        }
    }
```

**Comment by rayanbzd** [comment]
Exemple of a stacktrace. 
The stacktrace differ a bit depending on how complex the component is. But it's always about requireState on AdventureNBTSerialization
`[14:18:06 WARN] [com.github.retrooper.packetevents.PacketEventsAPI]: PacketEvents caught an unhandled exception while calling your listener.
java.lang.IllegalStateException: null
        at com.github.retrooper.packetevents.util.adventure.AdventureNBTSerialization.requireState(AdventureNBTSerialization.java:151) ~[?:?]
        at com.github.retrooper.packetevents.util.adventure.AdventureNBTSerialization.readStyle(AdventureNBTSerialization.java:306) ~[?:?]
        at com.github.retrooper.packetevents.util.adventure.AdventureNBTSerialization.readComponent(AdventureNBTSerialization.java:739) ~[?:?]
        at com.github.retrooper.packetevents.util.adventure.AdventureNBTSerialization.readComponent(AdventureNBTSerialization.java:604) ~[?:?]
        at com.github.retrooper.packetevents.util.adventure.AdventureNBTSerialization.readComponent(AdventureNBTSerialization.java:600) ~[?:?]
        at com.github.retrooper.packetevents.util.adventure.AdventureNBTSerialization.readComponent(AdventureNBTSerialization.java:595) ~[?:?]
        at com.github.retrooper.packetevents.wrapper.PacketWrapper.readComponentAsNBT(PacketWrapper.java:529) ~[?:?]
        at com.github.retrooper.packetevents.wrapper.PacketWrapper.readComponent(PacketWrapper.java:524) ~[?:?]
        at com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerSystemChatMessage.read(WrapperPlayServerSystemChatMessage.java:90) ~[?:?]
        at com.github.retrooper.packetevents.wrapper.PacketWrapper.readEvent(PacketWrapper.java:238) ~[?:?]
        at com.github.retrooper.packetevents.wrapper.PacketWrapper.<init>(PacketWrapper.java:143) ~[?:?]
        at com.github.retrooper.packetevents.wrapper.PacketWrapper.<init>(PacketWrapper.java:133) ~[?:?]
        at com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerSystemChatMessage.<init>(WrapperPlayServerSystemChatMessage.java:50) ~[?:?]
        at net.zencraft.velocity.chathistory.SystemChatPacketListener.onPacketSend(SystemChatPacketListener.java:26) ~[?:?]
        at com.github.retrooper.packetevents.event.PacketSendEvent.call(PacketSendEvent.java:51) ~[?:?]
        at com.github.retrooper.packetevents.event.EventManager.callEvent(EventManager.java:52) ~[?:?]
        at io.github.retrooper.packetevents.handlers.PacketEventsEncoder.read(PacketEventsEncoder.java:46) ~[?:?]
        at io.github.retrooper.packetevents.handlers.PacketEventsEncoder.encode(PacketEventsEncoder.java:68) ~[?:?]
        at io.github.retrooper.packetevents.handlers.PacketEventsEncoder.encode(PacketEventsEncoder.java:32) ~[?:?]
        at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:107) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:881) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:863) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:968) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:856) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:113) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:881) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:863) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:968) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:856) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.ChannelDuplexHandler.write(ChannelDuplexHandler.java:115) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at net.william278.velocitab.packet.PlayerChannelHandler.write(PlayerChannelHandler.java:43) ~[?:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:879) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:940) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1247) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:413) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[velocity.jar:3.3.0-SNAPSHOT (git-59567512-b346)]
        at java.lang.Thread.run(Thread.java:840) [?:?]`

**Comment by rayanbzd** [comment]
I tried to use directly the the method ComponentHolder.read() from velocity and it work properly without any errors, even with complex components. 
So it confirm that there is a problem using AdventureNBTSerialization of packetevents
```
    @Override
    public void onPacketSend(PacketSendEvent event) {
        Player player = (Player) event.getPlayer();
        if(event.getConnectionState() != ConnectionState.PLAY)return;
        if(event.getUser().getClientVersion().isOlderThan(ClientVersion.V_1_20_2))return;
        if (event.getPacketType() == PacketType.Play.Server.SYSTEM_CHAT_MESSAGE) {
            ComponentHolder component = ComponentHolder.read((ByteBuf) event.getByteBuf(), ProtocolVersion.MINECRAFT_1_20_3);
            ChatHistory.getInstance().addMessage(player, component.getComponent());
        }
    }
```

**Comment by booky10** [comment]
Which Component are you sending to the player?

The stacktrace indicates the hover event action being sent after its contents, which currently require the action to be sent first.

Should be pretty easy to fix by determining the contents type by the tag type instead of requiring the action to be sent first.

**Comment by rayanbzd** [comment]
> Which Component are you sending to the player?
> 
> The stacktrace indicates the hover event action being sent after its contents, which currently require the action to be sent first.
> 
> Should be pretty easy to fix by determining the contents type by the tag type instead of requiring the action to be sent first.

Since 1.20.2, chat is cleared on client side while changing backend server. I'm actually creating a plugin that store in cache the chat of the player to send it back to him when he change backend server. 

You can easily reproduce the problem by simply storing the default join message Component, then try to send it back to the player. 
![image](https://github.com/retrooper/packetevents/assets/52638579/d830b9c7-eb07-414b-af98-6df77b3425e7)

**Comment by IbanEtchep** [comment]
I have the same issue in Bungeecord with every message i try to capture and send back to the player :/

```java
        if (event.getPacketId() == PacketType.Play.Server.SYSTEM_CHAT_MESSAGE.getId(user.getClientVersion())) {
            WrapperPlayServerSystemChatMessage wrapper = new WrapperPlayServerSystemChatMessage(event);
            Component message = wrapper.getMessage();
//            plugin.getAdventure().player(player).sendMessage(message);
            BaseComponent[] baseComponents = BungeeComponentSerializer.get().serialize(message);
            player.sendMessage(baseComponents);
        }
```

`[19:42:52 WARN] [fr.iban.bungeechatkeeper.packetevents.api.PacketEventsAPI]: PacketEvents caught an unhandled exception while calling your listener.
java.lang.IllegalStateException: Illegal component nbt key read: 'type'
        at fr.iban.bungeechatkeeper.packetevents.api.util.adventure.AdventureNBTSerialization.readStyle(AdventureNBTSerialization.java:394) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.util.adventure.AdventureNBTSerialization.readComponent(AdventureNBTSerialization.java:739) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.util.adventure.AdventureNBTSerialization.readComponentList(AdventureNBTSerialization.java:952) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.util.adventure.AdventureNBTSerialization.readComponent(AdventureNBTSerialization.java:728) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.util.adventure.AdventureNBTSerialization.readComponent(AdventureNBTSerialization.java:604) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.util.adventure.AdventureNBTSerialization.readComponent(AdventureNBTSerialization.java:600) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.util.adventure.AdventureNBTSerialization.readComponent(AdventureNBTSerialization.java:595) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.wrapper.PacketWrapper.readComponentAsNBT(PacketWrapper.java:529) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.wrapper.PacketWrapper.readComponent(PacketWrapper.java:524) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.wrapper.play.server.WrapperPlayServerSystemChatMessage.read(WrapperPlayServerSystemChatMessage.java:90) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.wrapper.PacketWrapper.readEvent(PacketWrapper.java:238) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.wrapper.PacketWrapper.<init>(PacketWrapper.java:143) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.wrapper.PacketWrapper.<init>(PacketWrapper.java:133) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.wrapper.play.server.WrapperPlayServerSystemChatMessage.<init>(WrapperPlayServerSystemChatMessage.java:50) ~[?:?]
        at fr.iban.bungeechatkeeper.PacketEventsListener.onPacketPlaySend(PacketEventsListener.java:33) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.event.SimplePacketListenerAbstract.onPacketSend(SimplePacketListenerAbstract.java:64) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.event.PacketSendEvent.call(PacketSendEvent.java:51) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.api.event.EventManager.callEvent(EventManager.java:52) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.impl.handlers.PacketEventsEncoder.read(PacketEventsEncoder.java:53) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.impl.handlers.PacketEventsEncoder.encode(PacketEventsEncoder.java:79) ~[?:?]
        at fr.iban.bungeechatkeeper.packetevents.impl.handlers.PacketEventsEncoder.encode(PacketEventsEncoder.java:37) ~[?:?]
        at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:107) ~[waterfall.jar:git:Waterfall-Bootstrap:1.20-R0.2-SNAPSHOT:4a3846a:559]`

**Comment by AbhigyaKrishna** [comment]
Fixed in #683
